Introduce new constructors for flow::resource_limiter (#2137)
oneapi-src/oneTBB: rfcs/proposed/flow_graph_serializers/resource_consumer_node_api.md

how and when access will be granted to each node that requests it. The concurrency held by the currently processed input message is not 
released until all necessary resource accesses are granted and the body is executed.

### Extending the ``resource_limiter`` Constructors

The initial version of the resource limiting feature in the Flow Graph supported a single constructor for ``tbb::flow::resource_limiter`` class,
accepting a variadic number of resources:

```cpp
template <typename ResourceHandle>
class resource_limiter {
public:
    template <typename Handle, typename... Handles>
    resource_limiter(Handle&& handle, Handles&&... handles) // Constructor (1)
    {}
};

resource_limiter<int> limiter(1, 2, 3, 4, 5);
```

In the code above, ``limiter`` manages 5 resources of type ``int``. Instances of resources are constructed in-place while constructing the limiter.
The concern with this constructor is that the number of resources should be known at compile-time, while some workflows only knows it in runtime.

> Current underneath implementation of ``resource_limiter`` uses ``std::forward_list`` container which already provides a necessary runtime semantics, so the

Reviewer: ```suggestion
The current implementation of ``resource_limiter`` uses ``std::forward_list`` container which already provides the necessary runtime semantics, so the
```

only concern is the ``resource_limiter`` constructor itself.

> The idea is to extend the set of constructors to accept arbitrary containers (or ranges) to define the number of resources and its values in runtime.

Reviewer: ```suggestion
The idea is to extend the set of constructors to accept arbitrary containers (or ranges) to define the number of resources and their values at runtime.
```

```cpp
template <typename ResourceHandle>
class resource_limiter {
public:
    template <typename Handle, typename... Handles>
    resource_limiter(Handle&& handle, Handles&&... handles) // Constructor (1) - existing
    {}

    template <typename InputIterator>
    resource_limiter(InputIterator first, InputIterator last) // Constructor (2) - extension
    {}

    template <typename ContainerBasedSequence>

>     resource_limiter(ContainerBasedSequence&& sequence) // Constructor (3) - extension

Reviewer: Should we force failure on construction with an empty container / sequence? Since the resource_limited_node uses reserve_wait / release_wait pairs (in the current implementation), an empty container would lead to a call to graph::wait_for_all() being stuck. We can either document that such a thing is dangerous, assert or something else.

Author: Yes, but the only available way for the current API is to force it in runtime by adding an assert, exception or something like that.

Another option is to have 1 resource handle explicitly and others as a sequence, but I would prefer the existing API with an assert in debug/UB in release.

I have added the paragraph into the RFC about that.
